[BEAM-6522] Use fastavro as default Avro implementation for Avro source/sink in Python3

fredo838 opened the pull request:
This is is part of a series of PRs with goal to make Apache Beam PY3 compatible. The proposal with the outlined approach has been documented here: https://s.apache.org/beam-python-3.

This PR fixes support for Avro by making fastavro the default Avro implementation for Python3. Python2 continues to use avro library as the default.

CC: @Juta @RobbeSneyders 

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

RobbeSneyders commented:
Precommit fails on fastavro_it_test, which compares results of Avro and fastavro. With Avro deprecated on Python 3, this test doesn't make any sense. Let's skip this test on Python 3 and add a write/read test which checks the read results against the original instead.

RobbeSneyders commented:
The `use_fastavro` flag currently defaults to `False`. Since fastavro would be the only option on Python 3, this flag should probably default to True, and be ingored or throw an error when False on Python 3.
Changing this flag would also impact Python 2, but looking at the documentation, this doesn't seem like a problem?
https://github.com/apache/beam/blob/7722dd14f50b84287301533ffeb1790f5e395161/sdks/python/apache_beam/io/avroio.py#L146-L147

fredo838 commented:
Run Python PostCommit

fredo838 commented:
Run Python PreCommit

fredo838 commented:
@tvalentyn

Juta commented:
Run Python PostCommit

fredo838 commented:
Run Python PostCommit

tvalentyn commented on sdks/python/apache_beam/io/avroio_test.py:
```diff
@@ -21,22 +21,28 @@
 # pylint: disable=wrong-import-order, wrong-import-position, ungrouped-imports
 try:
-  from avro.schema import Parse # avro-python3 library for python3
+  from avro.schema import parse # avro library for python2
```
Per high-level comment, let's keep Py2 and Py3 Avro codepaths for now.  Previous version on the import may better alight with this requirement.

fredo838 commented:
Run Python PostCommit

fredo838 commented:
Run Python PostCommit

tvalentyn commented on sdks/python/apache_beam/io/avroio_test.py:
```diff
@@ -472,10 +419,147 @@ def test_sink_transform_snappy(self):
-class TestFastAvro(TestAvro):
+@unittest.skipIf(sys.version_info[0] == 3 and
+                 os.environ.get('RUN_SKIPPED_PY3_TESTS') != '1',
+                 'Due to a known issue in avro-python3 package, it '
```
Let's change this to a typical warning we have, such as:
```
                   'This test still needs to be fixed on Python 3. '
                   'See BEAM-6522.').
```

tvalentyn commented on sdks/python/apache_beam/io/avroio_test.py:
```diff
@@ -21,22 +21,27 @@
 import tempfile
 import unittest
 from builtins import range
+import sys
```
I had a brief look at the test code -  there are a few `@unittest.skip("Skip during debugging")`, are you still working on this file, @fredo838?
Overall the test logic seems a little fragmented across 3 classes, and I think we could restructure the test for easier readability & maintenance. I'd have to take a closer look to have concrete suggestions for improvement. @RobbeSneyders if you have time, please take a look to reduce the review feedback loop since you are in the same timezone.

tvalentyn commented on sdks/python/apache_beam/io/avroio.py:
```diff
@@ -70,15 +72,24 @@
 
 
+def _use_fastavro():
+  if sys.version_info[0] >= 3:
```
Also let's add a comment here with the reasoning and add a link to email thread on dev@.

tvalentyn commented:
Hi @fredo838, I didn't realize there were several changes on the PR recently. Is it ready for review? Once you finished the iteration of changes and would like reviewers to take another look you could   resolve threads that are resolved since previous review, e.g. comment "Done" on the threads, and give reviewers a heads-up comment on the PR, such as "PTAL".

fredo838 commented:
I addressed the comments you gave last week, but the code is still fragmented over multiple classes, so this is why I didn't notify anyone yet. This is planned for tomorrow.

fredo838 commented:
PTAL: @Juta

fredo838 commented:
PTAL: @tvalentyn

tvalentyn commented on sdks/python/apache_beam/io/avroio_test.py:
```diff
@@ -281,30 +277,34 @@ def test_read_with_splitting_multiple_blocks(self):
     num_records = 12000
-    file_name = self._write_data(count=num_records)
+    if self.use_fastavro:
+      sync_interval = 16000
```
Let's not pass sync_interval here, so that the default value is used, see also a comment below on this.

tvalentyn replied:
Discussed offline - the test code relies on sync_interval.

tvalentyn commented on sdks/python/apache_beam/io/avroio_test.py:
```diff
@@ -472,10 +476,55 @@ def test_sink_transform_snappy(self):
+                  prefix=tempfile.template,
+                  codec='null',
+                  count=len(RECORDS),
+                  sync_interval=16000):
```
Let's not redefine default values for `writer()` method in the test. Currently the defaults for `codec` and `sync_interval` match the respective defaults in `fastavro/avro`, but if the library implementation changes the default in the future, the test may be still hardcoding old value. How about something like:
```
  def _write_data(self,
                  directory=None,
                  prefix=tempfile.template,
                  count=len(RECORDS),
                  **kwargs):
    ...
    writer(f, self.SCHEMA, all_records, **kwargs)
```

tvalentyn commented on sdks/python/apache_beam/io/avroio.py:
```diff
@@ -562,9 +577,8 @@ def _create_avro_sink(file_path_prefix,
-  """A sink for avro files."""
-
+class _BaseAvroSink(filebasedsink.FileBasedSink):
+  """A FileBasedSink """
```
Let's drop or clarify this docstring.

tvalentyn commented:
LGTM, thanks @fredo838.

tvalentyn commented:
@chamikaramj Could you PTAL and merge this?

chamikaramj commented on sdks/python/apache_beam/examples/fastavro_it_test.py:
```diff
@@ -84,13 +84,11 @@ def record(i):
-                 os.environ.get('RUN_SKIPPED_PY3_TESTS') != '1',
-                 'This test still needs to be fixed on Python 3'
-                 'TODO: BEAM-6522')
+@unittest.skip('Due to a known issue in avro-python3 package, this'
```
Can we keep running this tests for Python 2 ? Otherwise this will become a regression in test coverage.

tvalentyn replied:
Thanks for catching this, @chamikaramj. It's good to have a fresh pair of eyes on the PR. Note that in an earlier revision this test was skipped only on Python-3  https://github.com/apache/beam/commit/553cc73252c4ecf81f9a759e63d6d794753c57b3.
This is another argument in favor of NOT force-pushing squashed commits after a review iteration, but doing a squash before a final merge after a PR is approved.

chamikaramj commented on sdks/python/apache_beam/io/avroio.py:
```diff
@@ -146,6 +153,10 @@ def __init__(self, file_pattern=None, min_bundle_size=0, validate=True,
     """
+    if sys.version_info[0] >= 3 and not use_fastavro:
+      warnings.warn("Due to a known issue in avro-python3 package, it is "
+                    "recommended to use fastavro with Beam Avro IO on "
```
Please prioritize fixing this so that this kinds of warnings can be removed (can be after submitting this PR).

tvalentyn replied:
AFAIK, the failure is fixed in Avro HEAD, and we are awaiting for Avro release.

tvalentyn replied:
Looks like Avro release might happen soon [1]: 
@fredo838, could you please check whether currenlty-disabled-on-python3 test passes if we install: https://dist.apache.org/repos/dist/dev/avro/avro-1.9.0-rc2/py3/avro-python3-1.9.0.tar.gz

You can pass that file as --extra_package to a Dataflow job.

Thank you.
[1] https://lists.apache.org/thread.html/ff724cff8fed12ebc8316e0dc1e876f0ca8080ab68b6165b7f109665@%3Cdev.avro.apache.org%3E

fredo838 replied:
I will take a look.

tvalentyn commented:
Run Python PostCommit

tvalentyn commented:
PTAL, @chamikaramj

chamikaramj commented:
LGTM. Thanks.
